src/core/tranche/quant.sol

```diff
@@ -17,79 +17,96 @@ pragma solidity >=0.4.24;
-        uint chi;
+        uint index;
         uint speed; // Accumulation per second
-        uint48 rho; // Last time the rate was accumulated
+        uint48 lastUpdated; // Last time the rate was accumulated
     }
     
-    Rate public iBorrow;
+    Rate public borrowRate;
     uint public debt;
+    uint public supplyRate;
+    bool public supplyRateFixed;
     
-    constructor() public {
+    constructor(address reserve_) public {
+        reserve = ReserveLike(reserve_);
         wards[msg.sender] = 1;
-        iBorrow.chi = ONE;
-        iBorrow.speed = ONE;
+        borrowRate.index = ONE;
+        borrowRate.speed = ONE;
+        supplyRate = ONE;
     }
 
     function file(bytes32 what, uint speed_) public note auth {
-         if (what == "iborrow") {
+         if (what ==  "borrowrate") {
             drip();
-            iBorrow.speed = speed_;
+            borrowRate.speed = speed_;
         }
-    }
-
-    function getSpeed() public returns(uint) {
-        return iBorrow.speed;
-    }
-
-    function UpdateIBorrow(uint supplySpeed, uint reserve) public note auth {
-        require (supplySpeed > 0);
-        if (now >= iBorrow.rho) {
+        else if (what ==  "supplyrate") {
             drip();
+            supplyRate = speed_;
         }
-        uint supplySpeed_ = sub(supplySpeed, ONE);
-        uint ratio = rdiv(add(reserve, debt), debt); 
-        iBorrow.speed = add(rmul(ratio, supplySpeed_), ONE);
     }
 
-    function drip() public note auth{
-        if (now >= iBorrow.rho) {
-            (uint latest, , uint wad) = compounding();
-            iBorrow.chi = latest;
-            iBorrow.rho = uint48(now);
-            debt = add(debt, wad);   
+    function setFixedSupplyRate(bool fixed_) public note auth {
+        supplyRateFixed = fixed_;
+    }
+
+    function updateBorrowRate() public note auth {
+        if (supplyRateFixed && supplyRate > 0) {
+            if (now >= borrowRate.lastUpdated) {
+                drip();
+            }
+            uint balance = reserve.balance();
+            uint supplyRate_ = sub(supplyRate, ONE);
+            uint ratio = rdiv(add(balance, debt), debt); 
+            borrowRate.speed = add(rmul(ratio, supplyRate_), ONE);
         }
     }
 
     function updateDebt(int wad) public note auth  {
-        if (now >= iBorrow.rho) {
+        if (now >= borrowRate.lastUpdated) {
             drip();
         }
         debt = uint(int(debt) + int(wad));
     }
 
+    function drip() internal {
+        if (now >= borrowRate.lastUpdated) {
+            (uint latest, , uint wad) = compounding();
+            borrowRate.index = latest;
+            borrowRate.lastUpdated = uint48(now);
+            debt = add(debt, wad);   
+        }
+    }
+
     function compounding() internal view returns (uint, uint, uint) {
-        uint48 rho = iBorrow.rho;
-        require(now >= rho);
-        uint speed = iBorrow.speed;
+        uint48 lastUpdated = borrowRate.lastUpdated;
```

Comment:
nitpick but i don't think this variable necessarily has to be declared

Reply:
I prefer to declare vars, if I am using them more then once to increase readability... no hard feelings, so let's agree on a convention here :)

Reply:
@ilinzweilin it is cheaper the other way :)